### PR TITLE
Introduction of support id

### LINKED ISSUE #100: [Support V2] Save trace Id along with jobs information
### What to do?

1. Save trace Id along with jobs information
(to be discussed) make additional index for trace id for status queries
2. Add a new field ( Trace ID = Support ID) to the sendTransactions relayer service
3. Add Trace Id to every log message along with Job Id ( if it was already created)

Part of big task for establishing well support processes – https://www.notion.so/blockscout/Establish-support-process-d93aeb68739d46808e17eee27263044c

Relates to:
-  https://github.com/zkBob/zkbob-ui/issues/94

### PR DESCRIPTION
**Pull request by LyzeOfKiel**
Changes:
* Add `traceId` support. Trace ID can be any string, including empty string. If any additional validation is required you can suggest in comments. Now, `/sendTransactions` method accepts only the following requests:
```
{
    transactions: PoolTx[] // same structure as before
    traceId: string
}
```
Closes #100 
* Remove `/sendTransaction` endpoint as it is not used in our client implementation anymore (@EvgenKor maybe we should double-check it).
* Remove `/transactions` endpoint as it is not used in our client implementation anymore.
* Fix spelling for `/limits` response. It also requires changes in client library.

### CODE REVIEW

File: zp-relayer/router.ts [hunk 1]
```diff
@@ -29,7 +29,6 @@ router.use((err: any, req: Request, res: Response, next: NextFunction) => {
 })
 
 router.get('/', endpoints.root)
-router.post('/sendTransaction', wrapErr(endpoints.sendTransaction))
 router.post('/sendTransactions', wrapErr(endpoints.sendTransactions))
 router.get('/transactions', wrapErr(endpoints.getTransactions))
```
Comment (k1rill-fedoseev) [inline zp-relayer/router.ts:33]: Should this be removed?

File: zp-relayer/workers/poolTxWorker.ts [hunk 2]
```diff
@@ -33,27 +33,28 @@ export async function createPoolTxWorker<T extends EstimationType>(
...
     }
 
-    const txs = job.data
+    const txs = job.data.transactions
+    const traceId = job.data.traceId
 
-    const logPrefix = `POOL WORKER: Job ${job.id}:`
-    logger.info('%s processing...', logPrefix)
-    logger.info('Received %s txs', txs.length)
+    const jobLogger = scopedLogger(`POOL WORKER: Job ${job.id}: `, { traceId })
```
Comment (k1rill-fedoseev) [inline zp-relayer/workers/poolTxWorker.ts:45]: I personally would prefer to have all the fields inside of the property object:
```
const workerLogger = logger.child({ worker: 'pool' })
...
const jobLogger = workerLogger.child({ jobId: job.id, traceId })
```
Wdyt?

File: zp-relayer/workers/poolTxWorker.ts [hunk 3]
```diff
@@ -78,7 +79,7 @@ export async function createPoolTxWorker<T extends EstimationType>(
 
         await updateNonce(++nonce)
 
-        logger.info(`${logPrefix} TX hash ${txHash}`)
+        jobLogger.info(`TX hash ${txHash}`)
```
Comment (k1rill-fedoseev) [inline zp-relayer/workers/poolTxWorker.ts:82]: I think it makes sense to log all tx-related logs with txHash in the properties object under a`txHash` key:
```
jobLogger.info('Sent tx', { txHash })
```

File: zp-relayer/router.ts (hunk 1 above)
Comment (LyzeOfKiel) [reply zp-relayer/router.ts:33]: Yes, this method is deprecated and not used by client library for a while. `/sendTransactions` basically does the same thing but can accept several transactions in one request.

File: zp-relayer/workers/poolTxWorker.ts (hunk 2 above)
Comment (LyzeOfKiel) [reply zp-relayer/workers/poolTxWorker.ts:45]: Agree, looks more concise. I will update this.

File: zp-relayer/router.ts (hunk 1 above)
Comment (k1rill-fedoseev) [reply zp-relayer/router.ts:33]: I meant this line:
```typescript
router.get('/transactions', wrapErr(endpoints.getTransactions))
```

Comment (LyzeOfKiel) [reply zp-relayer/router.ts:33]: Got it. Same thing applies to this endpoint as well. Not used by client any more.

Comment (LyzeOfKiel) [reply zp-relayer/router.ts:33]: Will remove it.

**Review by akolotov — COMMENTED** [review]
Consider to introduce
1/ TrackId for every endpoint so we can monitor activity of a user through all its actions,
2/ ability to turn on/off the requirement to have non-empty trackId through an environment variable

**Comment by LyzeOfKiel** [comment]
> Consider to introduce
> 1/ TrackId for every endpoint so we can monitor activity of a user through all its actions,
> 2/ ability to turn on/off the requirement to have non-empty trackId through an environment variable

@akolotov  In this way for `GET` requests `traceId` should be in query args (e.g. `/info?traceId=${traceId})` and for `POST` requests `traceId` should be in request body part?

**Comment by akolotov** [comment]
> @akolotov In this way for GET requests traceId should be in query args (e.g. /info?traceId=${traceId}) and for POST requests traceId should be in request body part?

Would it be better to have it as part of HTTP headers? Do you see limitations of this approach?

**Comment by LyzeOfKiel** [comment]
> > @akolotov In this way for GET requests traceId should be in query args (e.g. /info?traceId=${traceId}) and for POST requests traceId should be in request body part?
> 
> Would it be better to have it as part of HTTP headers? Do you see limitations of this approach?

Will test this approach.

File: zp-relayer/endpoints.ts [hunk 4]
```diff
@@ -9,20 +9,26 @@ import {
...
   if (errors) {
     logger.info('Request errors: %o', errors)
     res.status(400).json({ errors })
     return
   }
+  logger.info('TraceId %s', req.headers['trace-id'])
 
   const transactions = req.body.transactions as PoolTx[]
-  const traceId = req.body.traceId
+  const traceId = req.headers['trace-id'] as string
```
Comment (akolotov) [inline zp-relayer/endpoints.ts:31]: How about to obfuscate it a bit: `zkbob-support-id`?

**Comment by LyzeOfKiel** [comment]
@akolotov Added `trace-id` header for all endpoints except `/`, `/info`, `/params/hash/tree`, `/params/hash/tx`. I thought this would be too redundant. Should we add header validation for them as well? Or maybe we  should remove validation from some other endpoints?
Now, this header is logged for every endpoint if it passes validation. For `/sendTransactions` endpoint header is also logged several times during job processing.

**Comment by LyzeOfKiel** [comment]
Also, `RELAYER_REQUIRE_TRACE_ID` env can configure whether header validation is required or not.

**Comment by EvgenKor** [comment]
> Also, RELAYER_REQUIRE_TRACE_ID env can configure whether header validation is required or not.

In this way when RELAYER_REQUIRE_TRACE_ID is true we CAN add zkbob-support-id header. If the client didn't set this header, the relayer will no track it. Am I right?

**Comment by LyzeOfKiel** [comment]
> > Also, RELAYER_REQUIRE_TRACE_ID env can configure whether header validation is required or not.
> 
> In this way when RELAYER_REQUIRE_TRACE_ID is true we CAN add zkbob-support-id header. If the client didn't set this header, the relayer will no track it. Am I right?

If `RELAYER_REQUIRE_TRACE_ID` is `true` then relayer will require trace id header to be present, otherwise request will be rejected with code 400.
In other case, relayer will not validate the header. So, it doesn't matter whether it is present or not.

File: zp-relayer/validation/validation.ts [hunk 5]
```diff
@@ -77,39 +79,21 @@ const AjvSendTransactionSchema: JSONSchemaType<PoolTx> = {
...
-  type: 'array',
-  items: AjvSendTransactionSchema,
-}
-
-const AjvGetTransactionsSchema: JSONSchemaType<{
-  limit: number
-  offset: number
-  optimistic: boolean
-}> = {
+const AjvSendTransactionsSchema: JSONSchemaType<{ transactions: PoolTx[] }> = {
```
Comment (k1rill-fedoseev) [inline zp-relayer/validation/validation.ts:87]: Should we really change the schema here?

**Review by k1rill-fedoseev — APPROVED** [review]

File: zp-relayer/validation/validation.ts (hunk 5 above)
Comment (LyzeOfKiel) [reply zp-relayer/validation/validation.ts:87]: You mean that if don't use trace id as request body property any more then we can restore previously used function signature?

Comment (k1rill-fedoseev) [reply zp-relayer/validation/validation.ts:87]: Yeah

Comment (LyzeOfKiel) [reply zp-relayer/validation/validation.ts:87]: Agree, will restore previous signature

**Comment by LyzeOfKiel** [comment]
Updates:
* Now, `/sendTransactions` endpoint accepts only an array of transaction as before.
* Changed `trace-id` to `zkbob-support-id`.
* Refactored validation logic so that validation errors are now implicitly thrown and handled by express middleware.

File: zp-relayer/router.ts [hunk 6]
```diff
@@ -29,6 +31,13 @@ router.use((err: any, req: Request, res: Response, next: NextFunction) => {
   next()
 })
 
+router.use((req: Request, res: Response, next: NextFunction) => {
+  if (config.requireTraceId && req.headers[TRACE_ID]) {
+    logger.info('TraceId %s %s', req.headers[TRACE_ID], req.path)
```
Comment (k1rill-fedoseev) [inline zp-relayer/router.ts:36]: Let's put traceId and request path both under respective json keys in the log.

Http request from users are currently logged to a separate file, maybe it makes sense to put traceId there and log these messages to console as well.

Comment (LyzeOfKiel) [reply zp-relayer/router.ts:36]: >Let's put traceId and request path both under respective json keys in the log.

Agree

> maybe it makes sense to put traceId there and log these messages to console as well.

`zp.log` where every request is dumped, already includes all headers, so trace id will be saved there as well.
I don't think logging everything to console that is saved in `zp.log` is a good idea, as logs will be mostly flooded with such request logs which are >500 characters each

Comment (k1rill-fedoseev) [reply zp-relayer/router.ts:36]: > I don't think logging everything to console that is saved in zp.log is a good idea, as logs will be mostly flooded with such request logs which are >500 characters each

I think we do not need all of these 500 characters, but ok, let's leave it for future discussion/PR anyway.

**Comment by akolotov** [comment]
Will be merged as soon as https://github.com/zkBob/zeropool-relayer/pull/116 is closed.

**Review by EvgenKor — APPROVED** [review]

**Comment by akolotov** [comment]
@LyzeOfKiel one more thing - please update the README file with new information about the request header.

**Review by r0wdy1 — APPROVED** [review]

**Comment by akolotov** [comment]
@LyzeOfKiel sorry but another merge conflict appeared

**Pull request merged**